src/EditorFeatures/Core/Implementation/InlineRename/InlineRenameSession.cs

```diff
@@ -399,8 +405,27 @@ internal void ApplyReplacementText(string replacementText, bool propagateEditImm
                 }
 
                 _isApplyingEdit = false;
+
+                // We already kicked off UpdateConflictResolutionTask below (outside the delegate).
+                // Now that we are certain the replacement text has been propagated to all of the
+                // open buffers, it is safe to actually apply the replacements it has calculated.
+                // See https://devdiv.visualstudio.com/DevDiv/_workitems?_a=edit&id=227513
+                QueueApplyReplacements();
+
+                token.Dispose();
             };
 
+            // Start the conflict resolution task but do not apply the results immediately. The
+            // buffer changes performed in propagateEditAction can cause source control modal
+            // dialogs to show. Those dialogs pump, and yield the UI thread to whatever work is
+            // waiting to be done there, including our ApplyReplacements work. If ApplyReplacements
+            // starts running on the UI thread while propagateEditAction is still updating buffers
+            // on the UI thread, we crash because we try to enumerate the undo stack while an undo
+            // transaction is still in process. Therefore, we defer QueueApplyReplacements until
+            // after the buffers have been edited, and any modal dialogs have been completed.
```

Comment:
It might be worth mentioning that even if it wasn't for the undo case there's still a potential risk: we might end up applying the conflict stuff _first_, and then the non-conflict version second, leaving you in a temporarily stale state.